### DISCUSSION TITLE
Other kernels?

### DISCUSSION THREAD

**Discussion by untodesu** [fnuecke/oc2#224] [General]
I'm curious how the mod works in general, so...

Is it possible to use a different OS/kernel for the VM (say, [xv6](https://github.com/mit-pdos/xv6-riscv)) and if so, how can it be acheived?

> Also it seems like a nice idea for an open-source mod to have a custom OS, too bad people these days ain't interested in osdev...

**Comment by CoolCat467** [comment]
I believe [this](https://github.com/fnuecke/oc2/tree/1.18-forge/src/main/resources/data/oc2/file_systems) might be helpful

Edit: Ok that's not helpful for firmware. It's not currently possible it appears, I believe #127 is tracking this.

**Reply by untodesu** [reply]
It says only about additional data (ie mounting the hard drive inside already booted kernel) but nothing about booting (ish), if it's possible to boot a completely custom kernel then it's probably would be documented

**Reply by untodesu** [reply]
I see the JSON there mounting an archive (possibly produced at building) that in repo (probably) is represented by `src/main/scripts` and contains a bunch of Lua scripts in `bin` and some init.d (sysvinit?) scripts but no kernel binary.

Also the code contains a call to `Buildroot.getLinuxImage()`, so I assume the kernel image used by machines is hardcoded...

**Reply by untodesu** [reply]
The idea of custom kernels is interesting because people will be able to not just write applications for already running Linux kernel but also create operating systems designed specifically to run under OC2.

Also I, kind of, possibly, maybe, perhaps, wanted to rewrite Demos (and probably finish it lmao) to riscv and try it in Minecraft

**Reply by untodesu** [reply]
> Edit: Ok that's not helpful for firmware. It's not currently possible it appears, I believe #127 is tracking this.

Alright I see it!

**Comment by atirutw** [comment]
This will have to wait until they add custom firmware support since the kernel is in the firmware with everything else on a hard disk. This is why you can still try to boot and get a kernel panic even without a hard disk installed.

**Reply by untodesu** [reply]
I see, but still having _kernels_ in firmware feels a wee bit weird, it would rather be nicer to have some kind of a BIOS in the firmware that would load kernels from hard disks

**Reply by atirutw** [reply]
> having _kernels_ in firmware feels a wee bit weird

Yeah, but I guess it's easier to specifically boot Linux that way.

**Reply by untodesu** [reply]
> Yeah, but I guess it's easier to specifically boot Linux that way.

It also appears that RISC-V just boots systems that way... You just, kind of, write the kernel image to the board's memory (????).

**Reply by atirutw** [reply]
That's super weird. I guess they weren't thinking about custom OSes when writing standard firmware for RISC-V?

**Reply by untodesu** [reply]
At least HiFive1 appears to do that, the boot process differs for each existing RISC-V board

**Reply by untodesu** [reply]
> That's super weird. I guess they weren't thinking about custom OSes when writing standard firmware for RISC-V?

For firmware there's probably SBI and that's pretty much all, RISC-V literally has no BIOS or anything responsible for loading the kernel - that is either done by hardware or by some very proprietary ROM chips that mostly just read the memory where you'd write the kernel.

I am kind of beginning to understand the way RISCV works; so the kernel is _forced_ to provide some monolithic-ish drivers inside before it mounts any drives because we have no way to provide a ramdisk for the kernel (unlike x86 systems where the bootloader is responsible for loading ramdisks alongside kernel).
It's probably for the sake of simplifying the architecture, I can understand why they'd consider that instead of PC-like ways...

**Reply by fnuecke** [reply]
I think there's the theoretical possibility of going OpenSBI -> u-boot (with VirtIO drivers) -> kernel from disk, but I never got that to work (in the time I was willing to sink into it).

**Reply by untodesu** [reply]
> I think there's the theoretical possibility of going OpenSBI -> u-boot (with VirtIO drivers) -> kernel from disk, but I never got that to work (in the time I was willing to sink into it).

I just don't think having bootloaders on riscv is worth it, but custom firmware with kernels (referring to #127) would be a nice idea...

**Reply by atirutw** [reply]
I think it might be even better to create a general firmware that loads from disks similar how early x86 BIOS load OSes(load first 512 bytes, verify boot signature & JMP) and provide some basic IO stuff, but that's not for OC2 devs.

**Reply by untodesu** [reply]
> I think it might be even better to create a general firmware that loads from disks similar how early x86 BIOS load OSes(load first 512 bytes, verify boot signature & JMP) and provide some basic IO stuff, but that's not for OC2 devs.

I believe there are "proxy" kernels or whatever that do that very thing
